File: sdk/rust/src/runtime_provider.rs
use std::sync::Arc;
use std::time::SystemTime;

use tonic::codegen::async_trait;
use tonic::{Request as GrpcRequest, Response as GrpcResponse, Status};

use crate::agent::{AgentPreparedWorkspace, AgentWorkspace};
use crate::api::RuntimeMetadata;
use crate::error::Result as ProviderResult;
use crate::generated::v1::{self as pb};
use crate::protocol;
use crate::rpc_status::rpc_status;

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash)]
#[repr(i32)]
pub enum RuntimeEgressMode {
    #[default]
    Unspecified = 0,
    None = 1,
    Cidr = 2,
    Hostname = 3,
}

impl RuntimeEgressMode {
    pub const fn as_i32(self) -> i32 {
        self as i32
    }

    pub const fn from_i32_lossy(value: i32) -> Self {
        match value {
            1 => Self::None,
            2 => Self::Cidr,
            3 => Self::Hostname,
            _ => Self::Unspecified,
        }
    }
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct RuntimeSupport {
    pub can_host_apps: bool,
    pub egress_mode: RuntimeEgressMode,
    pub supports_prepare_workspace: bool,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct RuntimeSession {
    pub id: String,
    pub state: String,
    pub metadata: std::collections::BTreeMap<String, String>,
    pub lifecycle: Option<RuntimeSessionLifecycle>,
    pub state_reason: String,
    pub state_message: String,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct RuntimeSessionLifecycle {
    pub started_at: Option<SystemTime>,
    pub recommended_drain_at: Option<SystemTime>,
    pub expires_at: Option<SystemTime>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct RuntimeImagePullAuth {
    pub docker_config_json: String,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct StartRuntimeSessionRequest {
    pub app_name: String,
    pub template: String,
    pub image: String,
    pub metadata: std::collections::BTreeMap<String, String>,
    pub image_pull_auth: Option<RuntimeImagePullAuth>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct GetRuntimeSessionRequest {
    pub session_id: String,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ListRuntimeSessionsRequest {
    pub page_size: i32,
    pub page_token: String,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct ListRuntimeSessionsResponse {
    pub sessions: Vec<RuntimeSession>,
    pub next_page_token: String,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct StopRuntimeSessionRequest {
    pub session_id: String,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct PrepareRuntimeWorkspaceRequest {
    pub session_id: String,
    pub agent_session_id: String,
    pub workspace: Option<AgentWorkspace>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct PrepareRuntimeWorkspaceResponse {
    pub workspace: Option<AgentPreparedWorkspace>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct RemoveRuntimeWorkspaceRequest {
    pub session_id: String,
    pub agent_session_id: String,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct StartHostedAppRequest {
    pub session_id: String,
    pub app_name: String,
    pub command: String,
    pub args: Vec<String>,
    pub env: std::collections::BTreeMap<String, String>,
    pub allowed_hosts: Vec<String>,
    pub default_action: String,
    pub host_binary: String,
    pub workdir: String,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct HostedApp {
    pub id: String,
    pub session_id: String,
    pub app_name: String,
    pub dial_target: String,
}

fn support_to_proto(value: RuntimeSupport) -> pb::RuntimeSupport {
    pb::RuntimeSupport {
        can_host_apps: value.can_host_apps,
        egress_mode: value.egress_mode.as_i32(),
        supports_prepare_workspace: value.supports_prepare_workspace,
    }
}

fn session_to_proto(value: RuntimeSession) -> pb::RuntimeSession {
    pb::RuntimeSession {
        id: value.id,
        state: value.state,
        metadata: value.metadata,
        lifecycle: value
            .lifecycle
            .map(|lifecycle| pb::RuntimeSessionLifecycle {
                started_at: lifecycle
                    .started_at
                    .map(protocol::timestamp_from_system_time),
                recommended_drain_at: lifecycle
                    .recommended_drain_at
                    .map(protocol::timestamp_from_system_time),
                expires_at: lifecycle
                    .expires_at
                    .map(protocol::timestamp_from_system_time),
            }),
        state_reason: value.state_reason,
        state_message: value.state_message,
    }
}

fn start_session_request_from_proto(
    value: pb::StartRuntimeSessionRequest,
) -> StartRuntimeSessionRequest {
    StartRuntimeSessionRequest {
        app_name: value.app_name,
        template: value.template,
        image: value.image,
        metadata: value.metadata,
        image_pull_auth: value.image_pull_auth.map(|auth| RuntimeImagePullAuth {
            docker_config_json: auth.docker_config_json,
        }),
    }
}

fn list_sessions_response_to_proto(
    value: ListRuntimeSessionsResponse,
) -> pb::ListRuntimeSessionsResponse {
    pb::ListRuntimeSessionsResponse {
        sessions: value.sessions.into_iter().map(session_to_proto).collect(),
        next_page_token: value.next_page_token,
    }
}

fn list_sessions_request_from_proto(
    value: pb::ListRuntimeSessionsRequest,
) -> std::result::Result<ListRuntimeSessionsRequest, Status> {
    let mut page_size = value.page_size;
    if page_size < 0 {
        return Err(Status::invalid_argument("page_size must be non-negative"));
    }
    if page_size == 0 {
        page_size = 100;
    }
    if page_size > 200 {
        page_size = 200;
    }
    Ok(ListRuntimeSessionsRequest {
        page_size,
        page_token: value.page_token,
    })
}

fn prepare_workspace_request_from_proto(
    value: pb::PrepareRuntimeWorkspaceRequest,
) -> PrepareRuntimeWorkspaceRequest {
    PrepareRuntimeWorkspaceRequest {
        session_id: value.session_id,
        agent_session_id: value.agent_session_id,
        workspace: value.workspace.map(|workspace| AgentWorkspace {
            checkouts: workspace
                .checkouts
                .into_iter()
                .map(|checkout| crate::agent::AgentWorkspaceGitCheckout {
                    url: checkout.url,
                    reference: checkout.r#ref,
                    path: checkout.path,
                })
                .collect(),
            cwd: workspace.cwd,
        }),
    }
}

fn prepare_workspace_response_to_proto(
    value: PrepareRuntimeWorkspaceResponse,
) -> pb::PrepareRuntimeWorkspaceResponse {
    pb::PrepareRuntimeWorkspaceResponse {
        workspace: value.workspace.map(|workspace| pb::PreparedAgentWorkspace {
            root: workspace.root,
            cwd: workspace.cwd,
        }),
    }
}

fn start_app_request_from_proto(value: pb::StartHostedAppRequest) -> StartHostedAppRequest {
    StartHostedAppRequest {
        session_id: value.session_id,
        app_name: value.app_name,
        command: value.command,
        args: value.args,
        env: value.env,
        allowed_hosts: value.allowed_hosts,
        default_action: value.default_action,
        host_binary: value.host_binary,
        workdir: value.workdir,
    }
}

fn hosted_app_to_proto(value: HostedApp) -> pb::HostedApp {
    pb::HostedApp {
        id: value.id,
        session_id: value.session_id,
        app_name: value.app_name,
        dial_target: value.dial_target,
    }
}

#[async_trait]
/// Provider trait for serving hosted runtime sessions.
pub trait RuntimeProvider: Send + Sync + 'static {
    /// Configures the provider before it starts serving requests.
    async fn configure(
        &self,
        _name: &str,
        _config: serde_json::Map<String, serde_json::Value>,
    ) -> ProviderResult<()> {
        Ok(())
    }

    /// Returns runtime metadata that should augment the static manifest.
    fn metadata(&self) -> Option<RuntimeMetadata> {
        None
    }

    /// Returns non-fatal warnings the host should surface to users.
    fn warnings(&self) -> Vec<String> {
        Vec::new()
    }

    /// Performs an optional health check.
    async fn health_check(&self) -> ProviderResult<()> {
        Ok(())
    }

    /// Starts provider-owned background work after configuration.
    async fn start(&self) -> ProviderResult<()> {
        Ok(())
    }

    /// Shuts the provider down before the runtime exits.
    async fn close(&self) -> ProviderResult<()> {
        Ok(())
    }

    /// Returns the runtime capabilities supported by this provider.
    async fn get_support(&self, _request: ()) -> ProviderResult<RuntimeSupport> {
        Err(crate::Error::unimplemented(
            "runtime get support is not implemented",
        ))
    }

    /// Starts a hosted runtime session.
    async fn start_session(
        &self,
        _request: StartRuntimeSessionRequest,
    ) -> ProviderResult<RuntimeSession> {
        Err(crate::Error::unimplemented(
            "runtime start session is not implemented",
        ))
    }

    /// Returns one hosted runtime session by ID.
    async fn get_session(
        &self,
        _request: GetRuntimeSessionRequest,
    ) -> ProviderResult<RuntimeSession> {
        Err(crate::Error::unimplemented(
            "runtime get session is not implemented",
        ))
    }

    /// Lists hosted runtime sessions.
    async fn list_sessions(
        &self,
        _request: ListRuntimeSessionsRequest,
    ) -> ProviderResult<ListRuntimeSessionsResponse> {
        Err(crate::Error::unimplemented(
            "runtime list sessions is not implemented",
        ))
    }

    /// Stops a hosted runtime session.
    async fn stop_session(&self, _request: StopRuntimeSessionRequest) -> ProviderResult<()> {
        Err(crate::Error::unimplemented(
            "runtime stop session is not implemented",
        ))
    }

    /// Prepares an agent workspace for use by a hosted app.
    async fn prepare_workspace(
        &self,
        _request: PrepareRuntimeWorkspaceRequest,
    ) -> ProviderResult<PrepareRuntimeWorkspaceResponse> {
        Err(crate::Error::unimplemented(
            "runtime prepare workspace is not implemented",
        ))
    }

    /// Removes a previously prepared agent workspace.
    async fn remove_workspace(
        &self,
        _request: RemoveRuntimeWorkspaceRequest,
    ) -> ProviderResult<()> {
        Err(crate::Error::unimplemented(
            "runtime remove workspace is not implemented",
        ))
    }

    /// Starts one hosted app process inside a runtime session.
    async fn start_app(&self, _request: StartHostedAppRequest) -> ProviderResult<HostedApp> {
        Err(crate::Error::unimplemented(
            "runtime start app is not implemented",
        ))
    }
}

#[derive(Clone)]
pub(crate) struct RuntimeServer<P> {
    provider: Arc<P>,
}

impl<P> RuntimeServer<P> {
    pub(crate) fn new(provider: Arc<P>) -> Self {
        Self { provider }
    }
}

#[async_trait]
impl<P> pb::runtime_provider_server::RuntimeProvider for RuntimeServer<P>
where
    P: RuntimeProvider,
{
    async fn get_support(
        &self,
        request: GrpcRequest<()>,
    ) -> std::result::Result<GrpcResponse<pb::RuntimeSupport>, Status> {
        let support = self
            .provider
            .get_support(request.into_inner())
            .await
            .map_err(|error| rpc_status("runtime get support", error))?;
        Ok(GrpcResponse::new(support_to_proto(support)))
    }

    async fn start_session(
        &self,
        request: GrpcRequest<pb::StartRuntimeSessionRequest>,
    ) -> std::result::Result<GrpcResponse<pb::RuntimeSession>, Status> {
        let session = self
            .provider
            .start_session(start_session_request_from_proto(request.into_inner()))
            .await
            .map_err(|error| rpc_status("runtime start session", error))?;
        Ok(GrpcResponse::new(session_to_proto(session)))
    }

    async fn get_session(
        &self,
        request: GrpcRequest<pb::GetRuntimeSessionRequest>,
    ) -> std::result::Result<GrpcResponse<pb::RuntimeSession>, Status> {
        let session = self
            .provider
            .get_session({
                let request = request.into_inner();
                GetRuntimeSessionRequest {
                    session_id: request.session_id,
                }
            })
            .await
            .map_err(|error| rpc_status("runtime get session", error))?;
        Ok(GrpcResponse::new(session_to_proto(session)))
    }

    async fn list_sessions(
        &self,
        request: GrpcRequest<pb::ListRuntimeSessionsRequest>,
    ) -> std::result::Result<GrpcResponse<pb::ListRuntimeSessionsResponse>, Status> {
        let response = self
            .provider
            .list_sessions(list_sessions_request_from_proto(request.into_inner())?)
            .await
            .map_err(|error| rpc_status("runtime list sessions", error))?;
        Ok(GrpcResponse::new(list_sessions_response_to_proto(response)))
    }

    async fn stop_session(
        &self,
        request: GrpcRequest<pb::StopRuntimeSessionRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        self.provider
            .stop_session({
                let request = request.into_inner();
                StopRuntimeSessionRequest {
                    session_id: request.session_id,
                }
            })
            .await
            .map_err(|error| rpc_status("runtime stop session", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn prepare_workspace(
        &self,
        request: GrpcRequest<pb::PrepareRuntimeWorkspaceRequest>,
    ) -> std::result::Result<GrpcResponse<pb::PrepareRuntimeWorkspaceResponse>, Status> {
        let response = self
            .provider
            .prepare_workspace(prepare_workspace_request_from_proto(request.into_inner()))
            .await
            .map_err(|error| rpc_status("runtime prepare workspace", error))?;
        Ok(GrpcResponse::new(prepare_workspace_response_to_proto(
            response,
        )))
    }

    async fn remove_workspace(
        &self,
        request: GrpcRequest<pb::RemoveRuntimeWorkspaceRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        self.provider
            .remove_workspace({
                let request = request.into_inner();
                RemoveRuntimeWorkspaceRequest {
                    session_id: request.session_id,
                    agent_session_id: request.agent_session_id,
                }
            })
            .await
            .map_err(|error| rpc_status("runtime remove workspace", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn start_app(
        &self,
        request: GrpcRequest<pb::StartHostedAppRequest>,
    ) -> std::result::Result<GrpcResponse<pb::HostedApp>, Status> {
        let app = self
            .provider
            .start_app(start_app_request_from_proto(request.into_inner()))
            .await
            .map_err(|error| rpc_status("runtime start app", error))?;
        Ok(GrpcResponse::new(hosted_app_to_proto(app)))
    }
}

File: sdk/typescript/src/runtime-provider.ts
import { create } from "@bufbuild/protobuf";
import { EmptySchema } from "@bufbuild/protobuf/wkt";
import {
  Code,
  ConnectError,
  type ServiceImpl,
} from "@connectrpc/connect";

import {
  HostedAppSchema,
  ListRuntimeSessionsResponseSchema,
  PrepareRuntimeWorkspaceResponseSchema,
  RuntimeEgressMode as ProtoRuntimeEgressMode,
  Runtime as RuntimeProviderService,
  RuntimeSessionSchema,
  RuntimeSupportSchema,
  type GetRuntimeSessionRequest as ProtoGetRuntimeSessionRequest,
  type ListRuntimeSessionsRequest as ProtoListRuntimeSessionsRequest,
  type PrepareRuntimeWorkspaceRequest as ProtoPrepareRuntimeWorkspaceRequest,
  type RemoveRuntimeWorkspaceRequest as ProtoRemoveRuntimeWorkspaceRequest,
  type StartHostedAppRequest as ProtoStartHostedAppRequest,
  type StartRuntimeSessionRequest as ProtoStartRuntimeSessionRequest,
  type StopRuntimeSessionRequest as ProtoStopRuntimeSessionRequest,
} from "./internal/gen/v1/runtime_provider_pb.ts";
import {
  timestampFromDate,
} from "./protocol.ts";
import { errorMessage, type MaybePromise } from "./api.ts";
import { ProviderBase, type ProviderBaseOptions } from "./provider.ts";

type RuntimeProviderServiceImpl = Partial<
  ServiceImpl<typeof RuntimeProviderService>
>;

/** Native egress-mode constants for authored runtime providers. */
export const RuntimeEgressMode = {
  UNSPECIFIED: ProtoRuntimeEgressMode.UNSPECIFIED,
  NONE: ProtoRuntimeEgressMode.NONE,
  CIDR: ProtoRuntimeEgressMode.CIDR,
  HOSTNAME: ProtoRuntimeEgressMode.HOSTNAME,
} as const;
export type RuntimeEgressMode =
  (typeof RuntimeEgressMode)[keyof typeof RuntimeEgressMode];

export interface RuntimeSupport {
  canHostApps?: boolean | undefined;
  egressMode?: RuntimeEgressMode | undefined;
  supportsPrepareWorkspace?: boolean | undefined;
}

export interface RuntimeSessionLifecycle {
  startedAt?: Date | undefined;
  recommendedDrainAt?: Date | undefined;
  expiresAt?: Date | undefined;
}

export interface RuntimeSession {
  id?: string | undefined;
  state?: string | undefined;
  metadata?: Record<string, string> | undefined;
  lifecycle?: RuntimeSessionLifecycle | undefined;
  stateReason?: string | undefined;
  stateMessage?: string | undefined;
}

export interface RuntimeImagePullAuth {
  dockerConfigJson?: string | undefined;
}

export interface StartRuntimeSessionRequest {
  appName: string;
  template?: string | undefined;
  image?: string | undefined;
  metadata?: Record<string, string> | undefined;
  imagePullAuth?: RuntimeImagePullAuth | undefined;
}

export interface GetRuntimeSessionRequest {
  sessionId: string;
}

export interface ListRuntimeSessionsRequest {
  pageSize?: number | undefined;
  pageToken?: string | undefined;
}

export interface ListRuntimeSessionsResponse {
  sessions?: readonly RuntimeSession[] | undefined;
  nextPageToken?: string | undefined;
}

export interface StopRuntimeSessionRequest {
  sessionId: string;
}

export interface RuntimeAgentWorkspaceGitCheckout {
  url?: string | undefined;
  ref?: string | undefined;
  path?: string | undefined;
}

export interface RuntimeAgentWorkspace {
  checkouts?: readonly RuntimeAgentWorkspaceGitCheckout[] | undefined;
  cwd?: string | undefined;
}

export interface RuntimePreparedAgentWorkspace {
  root?: string | undefined;
  cwd?: string | undefined;
}

export interface PrepareRuntimeWorkspaceRequest {
  sessionId: string;
  agentSessionId: string;
  workspace?: RuntimeAgentWorkspace | undefined;
}

export interface PrepareRuntimeWorkspaceResponse {
  workspace?: RuntimePreparedAgentWorkspace | undefined;
}

export interface RemoveRuntimeWorkspaceRequest {
  sessionId: string;
  agentSessionId: string;
}

export interface StartHostedAppRequest {
  sessionId: string;
  appName: string;
  command?: string | undefined;
  args?: readonly string[] | undefined;
  env?: Record<string, string> | undefined;
  allowedHosts?: readonly string[] | undefined;
  defaultAction?: string | undefined;
  hostBinary?: string | undefined;
  workdir?: string | undefined;
}

export interface HostedApp {
  id?: string | undefined;
  sessionId?: string | undefined;
  appName?: string | undefined;
  dialTarget?: string | undefined;
}

export interface RuntimeProviderOptions extends ProviderBaseOptions {
  getSupport: () => MaybePromise<RuntimeSupport>;
  startSession: (
    request: StartRuntimeSessionRequest,
  ) => MaybePromise<RuntimeSession>;
  getSession: (
    request: GetRuntimeSessionRequest,
  ) => MaybePromise<RuntimeSession>;
  listSessions: (
    request: ListRuntimeSessionsRequest,
  ) => MaybePromise<ListRuntimeSessionsResponse>;
  stopSession: (request: StopRuntimeSessionRequest) => MaybePromise<void>;
  prepareWorkspace?: (
    request: PrepareRuntimeWorkspaceRequest,
  ) => MaybePromise<PrepareRuntimeWorkspaceResponse>;
  removeWorkspace?: (
    request: RemoveRuntimeWorkspaceRequest,
  ) => MaybePromise<void>;
  startApp: (
    request: StartHostedAppRequest,
  ) => MaybePromise<HostedApp>;
}

export class RuntimeProvider extends ProviderBase {
  readonly kind = "runtime" as const;

  private readonly getSupportHandler: RuntimeProviderOptions["getSupport"];
  private readonly startSessionHandler: RuntimeProviderOptions["startSession"];
  private readonly getSessionHandler: RuntimeProviderOptions["getSession"];
  private readonly listSessionsHandler: RuntimeProviderOptions["listSessions"];
  private readonly stopSessionHandler: RuntimeProviderOptions["stopSession"];
  private readonly prepareWorkspaceHandler: RuntimeProviderOptions["prepareWorkspace"];
  private readonly removeWorkspaceHandler: RuntimeProviderOptions["removeWorkspace"];
  private readonly startAppHandler: RuntimeProviderOptions["startApp"];

  constructor(options: RuntimeProviderOptions) {
    super(options);
    this.getSupportHandler = options.getSupport;
    this.startSessionHandler = options.startSession;
    this.getSessionHandler = options.getSession;
    this.listSessionsHandler = options.listSessions;
    this.stopSessionHandler = options.stopSession;
    this.prepareWorkspaceHandler = options.prepareWorkspace;
    this.removeWorkspaceHandler = options.removeWorkspace;
    this.startAppHandler = options.startApp;
  }

  async getSupport(): Promise<RuntimeSupport> {
    return await this.getSupportHandler();
  }

  async startSession(
    request: StartRuntimeSessionRequest,
  ): Promise<RuntimeSession> {
    return await this.startSessionHandler(request);
  }

  async getSession(
    request: GetRuntimeSessionRequest,
  ): Promise<RuntimeSession> {
    return await this.getSessionHandler(request);
  }

  async listSessions(
    request: ListRuntimeSessionsRequest,
  ): Promise<ListRuntimeSessionsResponse> {
    return await this.listSessionsHandler(request);
  }

  async stopSession(request: StopRuntimeSessionRequest): Promise<void> {
    await this.stopSessionHandler(request);
  }

  async prepareWorkspace(
    request: PrepareRuntimeWorkspaceRequest,
  ): Promise<PrepareRuntimeWorkspaceResponse> {
    if (!this.prepareWorkspaceHandler) {
      throw new ConnectError(
        "runtime provider prepare workspace is not implemented",
        Code.Unimplemented,
      );
    }
    return await this.prepareWorkspaceHandler(request);
  }

  async removeWorkspace(
    request: RemoveRuntimeWorkspaceRequest,
  ): Promise<void> {
    if (!this.removeWorkspaceHandler) {
      throw new ConnectError(
        "runtime provider remove workspace is not implemented",
        Code.Unimplemented,
      );
    }
    await this.removeWorkspaceHandler(request);
  }

  async startApp(
    request: StartHostedAppRequest,
  ): Promise<HostedApp> {
    return await this.startAppHandler(request);
  }
}

export function defineRuntimeProvider(
  options: RuntimeProviderOptions,
): RuntimeProvider {
  return new RuntimeProvider(options);
}

export function isRuntimeProvider(
  value: unknown,
): value is RuntimeProvider {
  return (
    value instanceof RuntimeProvider ||
    (typeof value === "object" &&
      value !== null &&
      "kind" in value &&
      (value as { kind?: unknown }).kind === "runtime" &&
      "getSupport" in value &&
      "startSession" in value &&
      "startApp" in value)
  );
}

export function createRuntimeProviderService(
  provider: RuntimeProvider,
): RuntimeProviderServiceImpl {
  return {
    async getSupport() {
      return create(
        RuntimeSupportSchema,
        runtimeSupportToProto(
          await invokeRuntimeProvider("get support", () =>
            provider.getSupport(),
          ),
        ),
      );
    },
    async startSession(request) {
      return create(
        RuntimeSessionSchema,
        runtimeSessionToProto(
          await invokeRuntimeProvider("start session", () =>
            provider.startSession(startSessionRequestFromProto(request)),
          ),
        ),
      );
    },
    async getSession(request) {
      return create(
        RuntimeSessionSchema,
        runtimeSessionToProto(
          await invokeRuntimeProvider("get session", () =>
            provider.getSession(getSessionRequestFromProto(request)),
          ),
        ),
      );
    },
    async listSessions(request) {
      const response = await invokeRuntimeProvider("list sessions", () =>
        provider.listSessions(listSessionsRequestFromProto(request)),
      );
      return create(ListRuntimeSessionsResponseSchema, {
        sessions: (response.sessions ?? []).map(runtimeSessionToProto),
        nextPageToken: response.nextPageToken ?? "",
      });
    },
    async stopSession(request) {
      await invokeRuntimeProvider("stop session", () =>
        provider.stopSession(stopSessionRequestFromProto(request)),
      );
      return create(EmptySchema);
    },
    async prepareWorkspace(request) {
      return create(
        PrepareRuntimeWorkspaceResponseSchema,
        prepareWorkspaceResponseToProto(
          await invokeRuntimeProvider("prepare workspace", () =>
            provider.prepareWorkspace(prepareWorkspaceRequestFromProto(request)),
          ),
        ),
      );
    },
    async removeWorkspace(request) {
      await invokeRuntimeProvider("remove workspace", () =>
        provider.removeWorkspace(removeWorkspaceRequestFromProto(request)),
      );
      return create(EmptySchema);
    },
    async startApp(request) {
      return create(
        HostedAppSchema,
        hostedAppToProto(
          await invokeRuntimeProvider("start app", () =>
            provider.startApp(startHostedAppRequestFromProto(request)),
          ),
        ),
      );
    },
  };
}

function startSessionRequestFromProto(
  request: ProtoStartRuntimeSessionRequest,
): StartRuntimeSessionRequest {
  return {
    appName: request.appName,
    template: request.template,
    image: request.image,
    metadata: { ...request.metadata },
    imagePullAuth: request.imagePullAuth === undefined
      ? undefined
      : { dockerConfigJson: request.imagePullAuth.dockerConfigJson },
  };
}

function getSessionRequestFromProto(
  request: ProtoGetRuntimeSessionRequest,
): GetRuntimeSessionRequest {
  return { sessionId: request.sessionId };
}

function listSessionsRequestFromProto(
  request: ProtoListRuntimeSessionsRequest,
): ListRuntimeSessionsRequest {
  const pageSize = request.pageSize;
  if (pageSize < 0) {
    throw new ConnectError(
      "page_size must be non-negative",
      Code.InvalidArgument,
    );
  }
  return {
    pageSize: pageSize === 0 ? 100 : Math.min(pageSize, 200),
    pageToken: request.pageToken,
  };
}

function stopSessionRequestFromProto(
  request: ProtoStopRuntimeSessionRequest,
): StopRuntimeSessionRequest {
  return { sessionId: request.sessionId };
}

function prepareWorkspaceRequestFromProto(
  request: ProtoPrepareRuntimeWorkspaceRequest,
): PrepareRuntimeWorkspaceRequest {
  return {
    sessionId: request.sessionId,
    agentSessionId: request.agentSessionId,
    workspace: request.workspace === undefined
      ? undefined
      : {
          checkouts: request.workspace.checkouts.map((checkout) => ({
            url: checkout.url,
            ref: checkout.ref,
            path: checkout.path,
          })),
          cwd: request.workspace.cwd,
        },
  };
}

function removeWorkspaceRequestFromProto(
  request: ProtoRemoveRuntimeWorkspaceRequest,
): RemoveRuntimeWorkspaceRequest {
  return {
    sessionId: request.sessionId,
    agentSessionId: request.agentSessionId,
  };
}

function startHostedAppRequestFromProto(
  request: ProtoStartHostedAppRequest,
): StartHostedAppRequest {
  return {
    sessionId: request.sessionId,
    appName: request.appName,
    command: request.command,
    args: [...request.args],
    env: { ...request.env },
    allowedHosts: [...request.allowedHosts],
    defaultAction: request.defaultAction,
    hostBinary: request.hostBinary,
    workdir: request.workdir,
  };
}

function runtimeSupportToProto(support: RuntimeSupport) {
  return {
    canHostApps: support.canHostApps ?? false,
    egressMode: support.egressMode ?? RuntimeEgressMode.UNSPECIFIED,
    supportsPrepareWorkspace: support.supportsPrepareWorkspace ?? false,
  };
}

function runtimeSessionToProto(session: RuntimeSession) {
  return {
    id: session.id ?? "",
    state: session.state ?? "",
    metadata: { ...(session.metadata ?? {}) },
    lifecycle: session.lifecycle === undefined
      ? undefined
      : {
          startedAt: session.lifecycle.startedAt === undefined
            ? undefined
            : timestampFromDate(session.lifecycle.startedAt),
          recommendedDrainAt: session.lifecycle.recommendedDrainAt === undefined
            ? undefined
            : timestampFromDate(session.lifecycle.recommendedDrainAt),
          expiresAt: session.lifecycle.expiresAt === undefined
            ? undefined
            : timestampFromDate(session.lifecycle.expiresAt),
        },
    stateReason: session.stateReason ?? "",
    stateMessage: session.stateMessage ?? "",
  };
}

function prepareWorkspaceResponseToProto(
  response: PrepareRuntimeWorkspaceResponse,
) {
  return {
    workspace: response.workspace === undefined
      ? undefined
      : {
          root: response.workspace.root ?? "",
          cwd: response.workspace.cwd ?? "",
        },
  };
}

function hostedAppToProto(app: HostedApp) {
  return {
    id: app.id ?? "",
    sessionId: app.sessionId ?? "",
    appName: app.appName ?? "",
    dialTarget: app.dialTarget ?? "",
  };
}

async function invokeRuntimeProvider<T>(
  label: string,
  fn: () => MaybePromise<T>,
): Promise<T> {
  try {
    return await fn();
  } catch (error) {
    if (error instanceof ConnectError) {
      throw error;
    }
    throw new ConnectError(
      `runtime provider ${label}: ${errorMessage(error)}`,
      Code.Unknown,
    );
  }
}
